Add skip-link to FSE themes

aristath opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/21507

## Description

This PR does the following:

* Adds a new `skipLinks` top-level item in `theme.json`
* Allows themes to opt-out of the auto-generated skip-link in theme.json
* Allows themes to define the skip-link target(s)
* Includes some sane fallback for elements that should be used as skip-link targets in case the author has not defined any targets, or the selector they defined was not found.

Then on render the following happen:

* We look inside the content HTML for things like `id="skip-link-target"` - assuming the user has defined `#skip-link-target` as the element in their theme.json file.
  * If the ID is located in the content, then we add the skip-link in `wp_body_open` and target the defined element.
  * If the ID is not located, then a vanilla-JS script gets injected in the footer and takes care of injecting the skip-link by querying the DOM for the fallback elements.

## `theme.json`

The `skipLinks` can be defined like this:
```json
{
	"skipLinks": {
		"auto": true,
		"css": true,
		"links": [
			{
				"target": "#skip-link-target",
				"label": "Skip to Content"
			},
			{
				"target": "#main-navigation",
				"label": "Skip to Navigation",
				"useFallbacks": false
			}
		]
	}
}
```
Arguments:
* `auto`: (bool) If set to false then automatic skip-links generation will be skipped. This is how authors can opt-out of the feature (for example if it's a single-page with no header and it goes straight to content :shrug:). If omitted then we assume `true`.
* `css:` (bool) We shouldn't force theme-authors to re-invent the wheel each time, so the CSS gets printed by default, but authors can opt-out if they want by setting it to `false`.
* `links`: (array) This is where users can define their skip-link (or multiple skip-links if needed). Each "link" can have the following:
  * `target`: (string|array) An element that should be used as the target.
  * `label`: (string) The skip-link's label. If omitted defaults to `Skip to content`.
  * `useFallbacks`: (bool) If set to false, then the fallbacks won't be added (more info about the fallbacks below). This only makes sense in case we have multiple skip-links defined.

## Fallback selectors
The fallback selectors are defined in an array:
```php
$fallbacks = array(
	'#skip-link-target',
	'main',
	'.wp-block-post-title',
	'.wp-block-query-loop',
	'.wp-block-post-content',
	'.entry-content',
	'h1',
	'h2',
);
```
This covers most (if not all) cases I could think of, and their priority makes sense but we can expand that logic and tweak them.

## Why not a separate block?

Adding a separate block for the skip-link and the skip-link target would introduce a lot of complexity for users. The UI would be cumbersome because it's basically invisible elements. Adding as a block would also allow people who don't necessarily understand the purpose or importance of a skip-link to point the skip-link to the wrong things (footer menus anyone?)
This automated implementation takes away all the guesswork, while still allowing users to point the skip-link to an element of their choosing if they want to, by adding the `skip-link-target` anchor to a block. It's a reasonable compromise between automation and freedom.

## How has this been tested?

Tested in an FSE theme:
* Index pages
* Archives
* Single posts
* Single pages

Tested both with a `#skip-link-target` element present, and without.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

aristath committed:
Add skip-link to FSE themes

aristath committed:
Add check for null selectors

aristath committed:
Allow opt-in to auto-css for the skip-link

aristath committed:
Escape things where needed

aristath committed:
Add inline comment

aristath committed:
This MUST NOT be escaped

TimothyBJacobs commented on lib/class-wp-theme-json.php:
```diff
@@ -1068,6 +1072,45 @@ public function get_custom_templates() {
+	 * @return string
+	 */
+	public function should_add_skip_link_styles() {
+		return isset( $this->theme_json['skipLink']['css'] ) && $this->theme_json['skipLink']['css'];
```
```suggestion
		return ! empty( $this->theme_json['skipLink']['css'] );
```

TimothyBJacobs commented on lib/full-site-editing/templates.php:
```diff
@@ -192,3 +192,88 @@ function set_unique_slug_on_create_template( $post_id ) {
+			?>
+			<script>
+			( function() {
+				var searchEl = [ '<?php echo implode( "','", $selectors ); // phpcs:ignore WordPress.Security.EscapeOutput ?>' ], contentEl, contentElID, parentEl, skipLink, i;
```
You could replace this with a `wp_json_encode` which might be a bit more readable.

aristath replied:
Good point. Fixed :+1:

TimothyBJacobs commented on lib/full-site-editing/templates.php:
```diff
@@ -192,3 +192,88 @@ function set_unique_slug_on_create_template( $post_id ) {
+	// Add a script to auto-generate the target and link based on the content we have.
+	add_action(
+		'wp_footer',
+		function() use ( $selectors ) {
```
I'm not sure if we care about this yet, but AFAIK using an anonymous function as a hook callback would be a violation of Core coding standards.

aristath replied:
Yeah, I was thinking about that too but the coding standard exists because we want users to be able to remove the hook.
In this case they can do it using `remove_action( 'wp_body_open', 'gutenberg_inject_skip_link' );` so I was thinking it's OK... I can see though why someone would want to remove just that one and not its parent.

Moved to a separate function in https://github.com/WordPress/gutenberg/pull/28946/commits/540a5cd108ec0ee9c62afeb7aac8a288527f3b71 :+1:

gziolo commented:
An interesting idea 👍🏻 I have some questions related to the proposal. What if someone wants to have more than 1 skip link? What if they want to have a different text for the link? I'm a bit worried that the simplistic format that `theme.json` promotes creates limitations that might be hard to overcome. Just to give an example, in WP-Admin there are two skip links:

<img width="257" alt="Screen Shot 2021-02-12 at 08 13 44" src="https://user-images.githubusercontent.com/699132/107740144-b513b400-6d0a-11eb-97c0-9a3beeb4b7d3.png">
<img width="218" alt="Screen Shot 2021-02-12 at 08 13 49" src="https://user-images.githubusercontent.com/699132/107740147-b5ac4a80-6d0a-11eb-8dc2-e396359f4d4c.png">

aristath committed:
Use wp_json_encode

aristath committed:
Don't use an anonymous function

aristath committed:
Allow multiple skip-links

aristath committed:
Update lib/class-wp-theme-json.php

Co-authored-by: Timothy Jacobs <timothy@ironbounddesigns.com>

aristath committed:
Merge remote-tracking branch 'origin/add/skip-link' into add/skip-link

aristath committed:
fix should_add_skip_link_styles check after the refactor

nosolosw commented:
Hey, I was looking at this for the code changes, but I believe is better that we clarify direction before diving into code.

According to https://github.com/WordPress/gutenberg/issues/21507#issuecomment-730703562 it looks like the direction we want to go is trying to absorb the concept of skip links into the template parts (header). I lack context on the specifics but I understand it can be a block or an attribute of the template part ― for which we need the ability to "understand" what's a header, which is under development judging by the status of https://github.com/WordPress/gutenberg/issues/26599 although a first PR to classify them has recently landed https://github.com/WordPress/gutenberg/pull/28410

carolinan commented:
The skip link needs to be the first focusable item on a page, no matter where there or if a header template part is used.

aristath commented:
> What if someone wants to have more than 1 skip link?

Interesting and valid! I refactored the implementation to allow for multiple skip-links and adjusted the PR description accordingly :+1:

aristath committed:
Merge branch 'master' into add/skip-link

carolinan commented:
Please include the documentation for the new feature.

aristath commented:
> Please include the documentation for the new feature.

If it gets to the point where it is agreed that we want to merge this then of course docs will be written. But not before then, it would be a waste of time :+1:

aristath committed:
make css opt-out

aristath committed:
classList.add can add multiple classes at once

carolinan commented:
Is there anything still blocking this?  I guess it needs two approvals?

aristath committed:
Merge branch 'trunk' into add/skip-link

aristath committed:
Add docs

aristath commented:
Added docs.

aristath commented:
cc @JustinyAhin for the docs in this PR 👍

JustinyAhin commented:
> cc @JustinyAhin for the docs in this PR 👍

Docs look good to me.
Thank you for writing them :)

carolinan commented:
I believe the docs covers everything.

aristath commented:
Though approved, I'd like some more feedback here before merging a change like this. It's important we get it right. 👍 
I'll leave this open for a while, give everyone the opportunity to voice their opinion.

gziolo commented on lib/full-site-editing/templates.php:
```diff
@@ -192,3 +192,118 @@ function set_unique_slug_on_create_template( $post_id ) {
+	// Add the skip-link styles if needed.
+	if ( WP_Theme_JSON_Resolver::get_theme_data()->should_add_skip_link_styles() ) {
+		echo '<style id="skip-link-styles">';
+		echo '.skip-link.screen-reader-text{border:0;clip:rect(1px,1px,1px,1px);clip-path:inset(50%);height:1px;margin:-1px;overflow:hidden;padding:0;position:absolute !important;width:1px;word-wrap:normal !important;}';
```
Isn't `screen-reader-text` already defined in WordPress or is it only for WP-Admin? See: https://make.wordpress.org/accessibility/2015/02/09/hiding-text-for-screen-readers-with-wordpress-core/

aristath replied:
Unfortunately, that's only for wp-admin... All themes add that same CSS manually in their CSS files. 🤷

aristath committed:
Merge branch 'trunk' into add/skip-link

aristath committed:
fix merge conflicts

aristath committed:
more merge-conflicts

aristath commented:
Converted to draft for now, after some discussion #30336 is simpler and more likely to land at this stage.

aristath commented:
Closing this one, #30336 was merged instead.
